src/lib.rs

```diff
@@ -0,0 +1,579 @@
+            name: "Templar Protocol NFT".to_string(),
+            symbol: "TEMPLAR".to_string(),
+            icon: None,
+            base_uri: None,
+            reference: None,
+            reference_hash: None,
+        };
+        Self {
+            vaults: LookupMap::new(StorageKey::Vaults),
+            nft_collections: UnorderedMap::new(StorageKey::NftCollections),
+            invites: LookupMap::new(StorageKey::Invites),
+            nft: NonFungibleToken::new(
+                StorageKey::NonFungibleToken,
+                env::current_account_id(),
+                Some(StorageKey::TokenMetadata),
+                Some(StorageKey::Enumeration),
+                Some(StorageKey::Approval),
+            ),
+            metadata: LazyOption::new(StorageKey::Metadata, Some(&metadata)),
+        }
+    }
+
+    #[payable]
+    pub fn create_vault(
+        &mut self,
+        nft_collection: AccountId,
+        collateral_asset: AccountId,
+        stablecoin: AccountId,
+        min_collateral_ratio: u64,
+    ) -> (String, TokenId) {
+        require!(
+            env::attached_deposit() >= NFT_MINT_FEE,
+            "Not enough deposit to create vault",
+        );
+
+        let (vault_id, vault) = Vault::new(collateral_asset, stablecoin, min_collateral_ratio);
+        require!(
+            !self.vaults.get(&vault_id).is_some(),
+            "Vault already exists",
+        );
+
+        self.vaults.insert(&vault_id, &vault);
+        self.nft_collections
+            .insert(&nft_collection.to_string(), &nft_collection);
+
+        // Mint an NFT for vault access
+        let token_id = (self.nft.nft_total_supply().0 + 1).to_string();
+        let token_metadata = TokenMetadata {
+            title: Some(format!("Vault Access NFT: {}", vault_id)),
+            description: Some("Grants access to a Templar Protocol vault".to_string()),
+            media: None,
+            media_hash: None,
+            copies: Some(1),
+            issued_at: None,
+            expires_at: None,
+            starts_at: None,
+            updated_at: None,
+            extra: None,
+            reference: None,
+            reference_hash: None,
+        };
+        self.nft.internal_mint(
+            token_id.clone(),
+            env::predecessor_account_id(),
+            Some(token_metadata),
+        );
+
+        (vault_id, token_id)
+    }
+
+    #[payable]
+    pub fn deposit_stablecoin(&mut self, vault_id: String, amount: U128) {
+        let nft_collection = self.get_nft_collection_for_vault(&vault_id);
+        require!(
+            self.owns_nft(&env::predecessor_account_id(), &nft_collection.to_string()),
+            "Not authorized",
+        );
+
+        // Instead of calling ft_transfer_call, we just log the intent
+        // The actual transfer should be initiated by the user
+        env::log_str(&format!(
+            "Deposit intent: {} tokens to vault {}",
+            amount.0, vault_id
+        ));
+    }
```

Comment:
Yes. There needs to be some sort of method to call for lenders to deposit stablecoins. I updated it in the latest commit. Think the loan test is failing because the `ft_transfer_call` initiates an XCC and I'm having trouble testing with that.